Add bootsnap to default Gemfile

burke opened the pull request:
This adds [bootsnap](https://github.com/Shopify/bootsnap) to newly-generated projects, which [generally reduces application boot times by over 50%](https://github.com/Shopify/bootsnap#quick-performance-overview).

---

I heard through the grapevine that people would be open to this.

Early releases of bootsnap only supported macOS, but recent versions now additionally support linux and (at least in theory) most *BSDs. Windows support is still not implemented.

This change is small, but not trivial, and it's been a long time since I've been in this codebase. I'm looking for guidance on how best to get this change documented, tested, and merged.

cc @rafaelfranca

burke committed:
Add bootsnap to default Gemfile:

Bootsnap precomputes load path resolution and caches ruby ISeq
and YAML parsing/compilation, reducing application boot time by
approximately 50% on supported configurations.

matthewd commented on railties/lib/rails/generators/rails/app/templates/config/boot.rb:
```diff
@@ -1,3 +1,17 @@
 
 require 'bundler/setup' # Set up gems listed in the Gemfile.
+
+if RUBY_ENGINE == 'ruby' && RUBY_PLATFORM =~ /darwin|linux|bsd/
```
We should always require bootsnap and let it decide whether it supports the current environment.

matthewd commented on railties/lib/rails/generators/rails/app/templates/Gemfile:
```diff
@@ -25,6 +25,10 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 <% if RUBY_ENGINE == 'ruby' -%>
+
+# Reduces boot times through caching; configured in config/boot.rb
+gem 'bootsnap', '~> 1.0.0', platforms: [:mri]
```
Can we loosen this version constraint a bit?

Can bootsnap be installed on all platforms, even if it won't Do Cool Stuff?

y-yagi replied:
bootsnap only supports Ruby 2.3.0 and higher, but Rails still supports Ruby 2.2. 
https://github.com/Shopify/bootsnap/blob/master/bootsnap.gemspec#L26 

Should we check the Ruby version?

matthewd replied:
@burke could bootsnap allow installation on 2.2, and just operate in a degraded mode? AIUI path caching would still work -- it's the iseq cache that is strictly 2.3+, right?

kaspth commented on railties/lib/rails/generators/rails/app/templates/config/boot.rb:
```diff
@@ -1,3 +1,17 @@
+    load_path_cache:      true,
+    autoload_paths_cache: true,
+    compile_cache_iseq:   true,
+    compile_cache_yaml:   true
```
Maybe we should follow bundlers lead:

```ruby
require "bundler/setup"    # Set up gems listed in the Gemfile
require "bootsnap/setup" # Caches and compiles gems for faster boot.
```

Which then just includes the above conventionalized config.

kaspth replied:
Could also be `require "bootsnap/railtie"` or Rails could just set up bootsnap automatically after executing `boot.rb` if the gem is present, e.g.:

```ruby
# Somewhere in the Rails bowels:
begin
  require "bootsnap"

  Bootsnap.setup(…)
rescue LoadError
end
```

yivo commented:
This is not necessary thing for most Rails apps. Things must be kept as simple as possible. In future this change may bring some bugs. If people need this it is better to create separate official gem `bootsnap-rails` which is officially tested and ready for Rails. Merging everything is a hell.

burke commented:
Ok -- I don't have any time this week to revise this, but here's what I'm going to do:

1. Get bootsnap to install correctly even on platforms where it can't use the native extension;
2. Have it understand which features can be enabled based on the current platform;
3. Add a `bootsnap/setup` requirable that sets up the maximal set of available features based on the environment, and assuming ActiveSupport will be in use.

The method used here will still be available, but bootsnap/setup will wrap it.

rafaelfranca commented:
@yivo we are not merging anything. It is not even being added as a dependency of a framework. It is being added to the generated Gemfile and if people don't want they can just remove. Rails give people sane defaults so they don't need to spend time thinking if they want to use something or not.

@burke seems like a good plan :+1:

burke commented:
I've updated this PR to use bootsnap-1.1.0.pre. I'll update again to 1.1.0 after some more testing/feedback.

Changes:

* Added a jruby version that doesn't include the native extension and doesn't support compile caching;
* `config/boot.rb` now just requires `bootsnap/setup`;
* `bootsnap/setup.rb` enables whichever features are available given the current environment;
* `bootsnap/setup.rb` attempts to derive the cache directory from the stack.

*See also [`bootsnap/setup.rb`](https://github.com/Shopify/bootsnap/blob/master/lib/bootsnap/setup.rb).*

Thoughts?

jeremy commented on railties/lib/rails/generators/rails/app/templates/Gemfile:
```diff
@@ -19,6 +19,9 @@ git_source(:github) { |repo| "https://github.com/#{repo}.git" }
 # gem 'capistrano-rails', group: :development
 
+# Reduces boot times through caching; required in config/boot.rb
+gem 'bootsnap', '~> 1.1.0.pre', require: false
```
Do we need a conservative `~>` dep? How about `>= 1.1.0.pre`

burke replied:
Sure, updated.

grosser commented:
hmmm I tried running with `compile_cache_iseq` and it broke code-coverage [single_cov](https://github.com/grosser/single_cov) ... so had to disable it in test https://github.com/zendesk/samson/pull/2038 ... so would be nice to see an easy escape hatch for this like an ENV var or `defined?(Coverage)` or so ...

burke commented:
Yeah -- it's unfortunately hard to detect if coverage is currently enabled. Ruby, perplexingly, doesn't provide an API for this. `defined?(Coverage)` is about as close as we can get. I'll change that.

burke commented:
Bootsnap bumped to [1.1.0](https://github.com/Shopify/bootsnap/blob/master/CHANGELOG.md#110)

guilleiguaran commented:
Not sure what can I be doing wrong but adding `bootsnap` to my app doubles the loading time:

```
~/src/app[master] % time bin/rails runner "puts Rails.env"
development
bin/rails runner "puts Rails.env"  10.34s user 3.74s system 56% cpu 24.864 total
```

```
~/src/app[master] % time bin/rails runner "puts Rails.env"
development
bin/rails runner "puts Rails.env"  7.22s user 1.69s system 17% cpu 51.722 total
```

I've tested this a dozen times getting similar numbers every time.

burke commented:
Are you running it in a VM? Is the application source in a shared folder? If so, set bootsnap's cache directory outside of your app, or symlink `tmp/cache` to some other location on disk.

guilleiguaran commented:
> Are you running it in a VM? Is the application source in a shared folder?

No and no. I tried removing the tmp/cache folder anyway and creating a symlink under /tmp but I don't have luck yet.

During the first run (when cache isn't created yet) the time is very similar to the normal boot time of the app:
```
~/src/cookpad[master] % time bin/rails runner "puts Rails.env"
development
bin/rails runner "puts Rails.env"  8.67s user 2.57s system 44% cpu 25.136 total
```

During subsequent boots the time goes up to 2x the normal time:
```
~/src/cookpad[master] % time bin/rails runner "puts Rails.env"
development
bin/rails runner "puts Rails.env"  6.42s user 1.42s system 15% cpu 49.072 total
```

burke commented:
Very strange, so it sounds like `RubyVM::InstructionSequence.load_from_binary` is running slower than loading the file directly from source.

What do you get back from:

```
ruby -v
uname -msr
```

burke commented:
It might make sense to move this to [an issue on bootsnap](https://github.com/Shopify/bootsnap/issues/new).

guilleiguaran commented:
@burke sure!! https://github.com/Shopify/bootsnap/issues/68

kaspth commented on Gemfile:
```diff
@@ -56,6 +56,9 @@ gem "libxml-ruby", platforms: :ruby
 gem "erubis", "~> 2.7.0", require: false
 
+# for railties app_generator_test
+gem "bootsnap", ">= 1.1.0", require: false
```
I don't like us needing to bundle Bootsnap in the main gemfile for our Railties tests to work. Can't we find out what Gemfile template / gems the Railties generated Rails app uses? @rafaelfranca are you familiar with gems and railties tests?

Or maybe we should just use `require "bootsnap/setup" rescue LoadError` in `config/boot.rb`.

rafaelfranca replied:
The application inside the railties test suite uses the Rails main Gemfile. I don't see a problem on adding bootsnap in the main gemfile, we already do to gems like jbuilder, erubis, etc.

kaspth reviewed (commented):
Just one more thing… plus squash your commits and add a changelog entry. Then I think we're good to go! 🤘

burke commented:
Rebased, squashed, and added changelog entry.

I'm not against searching for an alternative solution if the addition to the root Gemfile is unacceptable on balance, but I'm strongly against rescuing `LoadError` in every downstream project to support not adding a gem to the Rails development Gemfile.

kaspth commented:
I was just going to merge this, but @rafaelfranca's too fast for me! 😄

kirs commented:
🎉 🎉 🎉
